File: apps/ios/self-management/Features/Tasks/Domain/UseCases/CreateTaskUseCase.swift
//
//  CreateTaskUseCase.swift
//  self-management
//
//  LEARNING: Create Action Use Case (Tasks)
//

import Foundation

/// Use case handles creating a new task.
final class CreateTaskUseCase {

    private let repository: TaskRepositoryProtocol

    init(repository: TaskRepositoryProtocol) {
        self.repository = repository
    }

    /// Creates a new task for the currently authenticated user.
    /// - Parameters:
    ///   - title: The title of the task.
    ///   - description: An optional description.
    /// - Returns: The newly created TodoTask entity.
    func execute(title: String, description: String? = nil) async throws -> TodoTask {
        // Business Rule: A task cannot have an empty title.
        guard !title.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw NSError(domain: "CreateTaskUseCase", code: -1, userInfo: [NSLocalizedDescriptionKey: "Task title cannot be empty"])
        }
        
        return try await repository.createTask(title: title, description: description)
    }
}

File: apps/mobile/lib/features/tasks/domain/use_cases/create_task_use_case.dart
import '../entities/todo_task.dart';
import '../repositories/task_repository.dart';
import 'package:fpdart/fpdart.dart';
import '../../../../core/network/index.dart';

class CreateTaskUseCase {
  final TaskRepository _repository;

  CreateTaskUseCase(this._repository);

  Future<Either<Failure, TodoTask>> execute({
    required String title,
    String? description,
  }) async {
    if (title.trim().isEmpty) {
      return const Left(ValidationFailure('Task title cannot be empty'));
    }
    return await _repository.createTask(title: title, description: description);
  }
}
